File: frontend/src/modules/tickets/validators/BusinessRuleValidator.ts
/**
 * Chain of Responsibility node — validates client-side business rules.
 *
 * Responsibility (SRP): enforce only UI-level business rules before submitting.
 *     No text field checks, no file checks — only pre-submit domain constraints.
 * Depends on: BaseValidator (src/core/base/BaseValidator.ts).
 * Pattern: Chain of Responsibility node.
 * SOLID: SRP · OCP · LSP
 *
 * Rules enforced (client-side mirror of BE BusinessRuleValidator):
 *   - Submission only within business hours (Mon–Fri 07:00–20:00, local time)
 *     Note: the authoritative check lives in the BE; this is UX-only early feedback.
 *
 * OCP: new client-side rule = new node; this class unchanged.
 *
 * Note: duplicate-ticket check is NOT done here (requires API call) — it lives in BE only.
 */

import { BaseValidator, type ValidationResult } from '../../../core/base/BaseValidator'

/**
 * H#6 (audit fix): Business hours validator is now informational only.
 * Tickets can be created 24/7. If outside business hours, the form shows
 * a non-blocking warning: "Tu ticket será atendido en el próximo horario laboral."
 * This validator always returns isValid: true — the warning is handled by the UI.
 */
export class BusinessRuleValidator extends BaseValidator {
  static isBusinessHours(): boolean {
    const now = new Date()
    const day = now.getDay()   // 0=Sun, 6=Sat
    const hour = now.getHours()
    return day !== 0 && day !== 6 && hour >= 7 && hour < 20
  }

  validate(_data: unknown): ValidationResult {
    // Always valid — business hours are informational, not blocking
    return { isValid: true, field: '', errors: [] }
  }
}

File: backend/apps/tickets/validators/business_rule_validator.py
"""
Chain of Responsibility node — validates business rules for ticket creation.

Responsibility (SRP): enforce domain-level business rules only.
Depends on: BaseValidator (core/base/base_validator.py).
    Receives ticket_repository via __init__ for duplicate check (DIP).
Pattern: Chain of Responsibility node.
SOLID: SRP · OCP · DIP · LSP
"""

from __future__ import annotations

from core.base.base_validator import BaseValidator, ValidationResult

# Business hours: Monday–Friday 07:00–20:00 local time
_BUSINESS_START = 7
_BUSINESS_END = 20
_BUSINESS_DAYS = {0, 1, 2, 3, 4}  # Mon=0 … Fri=4


class BusinessRuleValidator(BaseValidator):
    """Validates business-hour constraint and duplicate-ticket rule."""

    def __init__(self, ticket_repository) -> None:
        super().__init__()
        self._ticket_repository = ticket_repository

    def validate(self, data: dict) -> ValidationResult:
        # ── Rule 1: business hours (INFORMATIONAL, not blocking) ──────────
        # Tickets can be created 24/7. If outside business hours, the response
        # includes a warning but the ticket IS created.
        # This is handled at the view layer, not here (H#6 from audit).

        # ── Rule 2: no active duplicate for same client + service + subject ─
        cliente_id = data.get("cliente_id")
        asunto = data.get("asunto", "")
        servicio_id = data.get("servicio_id")

        if cliente_id and asunto and servicio_id:
            duplicate = self._ticket_repository.find_active_duplicate(
                cliente_id=cliente_id,
                asunto=asunto,
                servicio_id=servicio_id,
            )
            if duplicate:
                return ValidationResult(
                    is_valid=False,
                    errors=["Ya existe un ticket activo con el mismo asunto y servicio."],
                    field_name="duplicado",
                )

        return ValidationResult(is_valid=True)
